### DISCUSSION TITLE
Setting the DHCP Hostname

### DISCUSSION THREAD

**Discussion by brokenbrigde9000** [micropython/micropython#11652] [ESP32]
I'm running v1.20.0 on an ESP32 and try to change the device name which shows up at the wifi router it connects to. I see a couple of posts that state you can use .config(dhcp_hostname='insert_hostname_here'. I use this sequence:

```
# cfg has the wifi credentials
ssid = cfg['wifi_ssid']
password = cfg['wifi_pw']
hname = cfg['device_id'] # identify the device with this hostname

station = network.WLAN(network.STA_IF)
station.active(True)
station.config(dhcp_hostname=hname)    
station.connect(ssid, password)
```
unfortunately, using dhcp_hostname=  gives me 'ValueError:'. Is this functionality not available for the ESP32? Maybe it is defined differently and I'm using the wrong parameters?

**Comment by jimmo** [comment]
Setting `dhcp_hostname` raises ValueError if the hostname is too long.

In v1.20 we made this parameter work the same across all ports (which imposes a maximum length of 15), but I think previously the ESP32 version allowed a longer hostname.

**Reply by brokenbrigde9000** [reply]
Perfect, that works like a charm! Is there any place I can find information like that written down? I could not find this in the docs.

**Comment by Josverl** [comment]
>  the same across all ports 

Does that include rp2 with w5500?

Happened to try that yesterday and it failed, I'll double check the host name for dns/dhcp compliance

**Reply by Josverl** [reply]
@Jimmo, The Wiznet 5500  on v1.20.0 does not recognize dhcp_hostname.
precomiled firmware: `MicroPython v1.20.0 on 2023-04-26; W5500-EVB-Pico with RP2040`

![image](https://github.com/micropython/micropython/assets/981654/664456d8-c391-41c7-bbfe-309ecb663259)

**Reply by glenn20** [reply]
Possibly related (re. hostname support for wiznet on ESP32): #11541.

**Reply by jimmo** [reply]
@Josverl I don't think this is supported on the Wiznet driver yet.

Also `network.hostname()` is (since v1.20) the preferred way to set the hostname (and will apply to all NIC types that support it).

**Reply by massimosala** [reply]
Hi Jim

I cannot set the hostname, neither with network.hostname nor dhcp_hostname in STA_IF.

Please see my last comment:
https://github.com/micropython/micropython/issues/11450

**Comment by ghost** [comment]
I did not know about 15 chars limit for the dhcp name. Is there other limitations? 

I just rewrote my asynchonous driver, which tries to connect to 2 predefined APs -> [GitHub] (https://github.com/divergentti/Micropython-for-ESP32/blob/main/esp32-bme680-oled/drivers/WIFICONN_AS.py) and perhaps I shall add more error checking to the code.

**Reply by liudr** [reply]
IIRC, as an ESP-IDF developer in my past life, this limit is 30 or 31 in ESP32. I had azure IoT device names in 32-char alphanumerical sequences that I had to shorten slightly to fit. I also found out while doing some tests. 15 chars are barely 2 words and a short number. What is the reason for shortening this? Can I change the port or the main mp to increase back to say 30 so I can compile custom firmware? Thanks.

**Reply by jimmo** [reply]
See `MICROPY_PY_NETWORK_HOSTNAME_MAX_LEN`

**Reply by liudr** [reply]
Thanks @jimmo 
I tried including MICROPY_PY_NETWORK_HOSTNAME_MAX_LEN=31 in sdkconfig.board and there was no effect.
I found a way to add compiler options in mpconfigboard.cmake as the following and it does take effect. I checked on my router with up to 30 chars when defining 31. I think the last byte is \0.

add_compile_definitions(MICROPY_PY_NETWORK_HOSTNAME_MAX_LEN=31)

Interesting tidbit:

This def is in modnetwork.h with an #ifndef and a default value of 16. Then in relevant files >=MICROPY_PY_NETWORK_HOSTNAME_MAX_LEN is considered valueError:

```
        if (len >= MICROPY_PY_NETWORK_HOSTNAME_MAX_LEN) {
            mp_raise_ValueError(NULL);
        }
```
So mp has a bunch of C function calls like mp_raise_ValueError() instead of C++ member functions?
